pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1139 @@
+    /// is available. If this returns ValueReconstructResult::Continue, look up
+    /// the predecessor layer and call again with the same 'reconstruct_data' to
+    /// collect more data.
+    pub(crate) async fn get_value_reconstruct_data(
+        &self,
+        key: Key,
+        lsn_range: Range<Lsn>,
+        reconstruct_data: &mut ValueReconstructState,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<ValueReconstructResult> {
+        use anyhow::ensure;
+
+        let layer = self.0.get_or_maybe_download(true, Some(ctx)).await?;
+        self.0
+            .access_stats
+            .record_access(LayerAccessKind::GetValueReconstructData, ctx);
+
+        if self.layer_desc().is_delta {
+            ensure!(lsn_range.start >= self.layer_desc().lsn_range.start);
+            ensure!(self.layer_desc().key_range.contains(&key));
+        } else {
+            ensure!(self.layer_desc().key_range.contains(&key));
+            ensure!(lsn_range.start >= self.layer_desc().image_layer_lsn());
+            ensure!(lsn_range.end >= self.layer_desc().image_layer_lsn());
+        }
+
+        layer
+            .get_value_reconstruct_data(key, lsn_range, reconstruct_data, &self.0)
+            .await
+    }
+
+    /// Download the layer if evicted.
+    ///
+    /// Will not error when it is already downloaded.
+    pub(crate) async fn get_or_download(&self) -> anyhow::Result<()> {
+        self.0.get_or_maybe_download(true, None).await?;
+        Ok(())
+    }
+
+    /// Creates a guard object which prohibit evicting this layer as long as the value is kept
+    /// around.
+    pub(crate) async fn guard_against_eviction(
+        &self,
+        allow_download: bool,
+    ) -> anyhow::Result<ResidentLayer> {
+        let downloaded = self.0.get_or_maybe_download(allow_download, None).await?;
+
+        Ok(ResidentLayer {
+            downloaded,
+            owner: self.clone(),
+        })
+    }
+
+    pub(crate) fn info(&self, reset: LayerAccessStatsReset) -> HistoricLayerInfo {
+        self.0.info(reset)
+    }
+
+    pub(crate) fn access_stats(&self) -> &LayerAccessStats {
+        &self.0.access_stats
+    }
+
+    pub(crate) fn local_path(&self) -> &Path {
+        &self.0.path
+    }
+
+    /// Traditional debug dumping facility
+    #[allow(unused)]
+    pub(crate) async fn dump(&self, verbose: bool, ctx: &RequestContext) -> anyhow::Result<()> {
+        self.0.desc.dump();
+
+        if verbose {
+            // for now, unconditionally download everything, even if that might not be wanted.
+            let l = self.0.get_or_maybe_download(true, Some(ctx)).await?;
+            l.dump(&self.0).await?
+        }
+
+        Ok(())
+    }
+}
+
+/// The download-ness ([`DownloadedLayer`]) can be either resident or wanted evicted.
+///
+/// However when we want something evicted, we cannot evict it right away as there might be current
+/// reads happening on it. It has been for example searched from [`LayerMap`] but not yet
```

Comment:
could you explain what "searched from" here means?

Reply:
`LayerMap::search` (and the obvious next step of `LayerManager::get_from_desc`).

Reply:
Maybe this could be clarified in the rustdocs somehow?

Reply:
9a714ac6b